### ISSUE TITLE
[4.1][Feature] New Operation - RespondToAjaxSelects

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/CRUD#2175]
Ok the name is pretty bad, totally accepting suggestions for a better one. But the gist of it is that, now that operations can have routes, we can make it a lot easier for devs to add Fields and Filters that load stuff with AJAX.

For example, for the [select_from_ajax field type](https://backpackforlaravel.com/docs/4.0/crud-fields#select2_from_ajax), instead of asking the dev to 
- create new routes for getting categories with AJAX
- create new controller for responding to AJAX

we could just ask him to 
- ```use RespondsToAjaxSelectsOperation;``` on your CategoryCrudController;

Should be fairly easy to create the operation. The only "tricky" thing would be to determine what attributes are searched by default. Food for thought.

**Comment by tabacitu** [comment]
The new ```indentifiableName()``` method (https://github.com/Laravel-Backpack/CRUD/issues/2050) would not be a solution for the above, since it's a method, not a column in the db. So it's not searchable.

**Comment by samuk190** [comment]
I suggest changing the name to AjaxSelectService

**Comment by tabacitu** [comment]
**Rephrasing the issue:** Since in v4 it's much easier to add routes and methods to a CRUD, using Operations, it looks like we can make the process of adding a [```select2_from_ajax``` or ```select2_from_ajax_multiple``` field](https://backpackforlaravel.com/docs/4.0/crud-fields#select2_from_ajax) MUCH easier by leveraging Operations.

Instead of asking people to do:

```php
Route::get('/api/category', 'Api\CategoryController@index');
Route::get('/api/category/{id}', 'Api\CategoryController@show');
```

and
```php
<?php

namespace App\Http\Controllers\Api;

use Illuminate\Http\Request;
use App\Http\Controllers\Controller;
use Backpack\NewsCRUD\app\Models\Category;

class CategoryController extends Controller
{
    public function index(Request $request)
    {
        $search_term = $request->input('q');
        $page = $request->input('page');

        if ($search_term)
        {
            $results = Category::where('name', 'LIKE', '%'.$search_term.'%')->paginate(10);
        }
        else
        {
            $results = Category::paginate(10);
        }

        return $results;
    }

    public function show($id)
    {
        return Category::find($id);
    }
}
```

We could include the routes and methods as an operation, so it's only one or two lines of code. Possible solutions below.

**Note:** In the examples, the ```MonsterCrudController``` has a ```select2_from_ajax``` that allows the admin to select entries from the ```Category``` model.

# Solution 1) Operation on the CategoryCrudController

We could ask them to do:
``` use \Backpack\CRUD\app\Http\Controllers\Operations\RespondToSelectFieldCalls;``` on ```CategoryCrudController```.

That file would have a trait that looks something like:

```php
trait RespondToSelectFieldCalls
{
     protected function setupRespondToSelectRoutes($segment, $routeName, $controller) {}
     protected function setupRespondToSelectDefaults() {}
     public function respondToSelectSearch(Request $request) {}
     public function respondToSelectItem($id) {}
}
```

One problem with this is that you need to specify which attributes are searchable through the select. We could do that using:
- an ```identifiableAttribute()``` on the Model, as mentioned in https://github.com/Laravel-Backpack/CRUD/issues/2050 ;
- additionally, a property on the CategoryCrudController:
```php
protected $searchableAttributesInSelectFieldCalls = ['name'];
```

But this does have two problems:
(P1) What if you have more than one CRUDs that make ajax calls to fetch a Category, but different things are searchable? So ```MonsterCrudController``` makes AJAX calls for Category, searching for ```name```, but also ```ArticleCrudController``` and ```TagCrudController``` make calls to fetch categories, and they want to search by ```name```, but also by ```id``` and ```description```. You can't do that using this method, since you have only one endpoint responding to requests.
(P2) What if you don't have a CRUD for the Category at all? Why wouldn't you be able to use a Model that has no CRUD as a data source for your selects?

# Solution 2) Operation on the MonsterCrudController

We could ask them to do:
``` use \Backpack\CRUD\app\Http\Controllers\Operations\Fetch;``` on ```MonsterCrudController``` (the controller where the field is placed).

That file would have a trait that looks something like:

```php
trait Fetch
{
     protected function setupFetchRoutes($segment, $routeName, $controller) {}
     protected function setupFetchDefaults() {}
     public function fetchMultipleItems(Request $request) {}
     public function fetchOneItem($id) {}
}
```

In order to specify which Models and which Attribute are searchable, we could do:
```php
protected $fetch = [
    'category' => [
        'model' => Backpack\NewsCRUD\app\Models\Category::class, // required
        'routeSegment' => 'fetch/category', // falls back to the key of this array if not specified ("category")
        'searchableAttributes' => ['name'], // falls back to identifiableAttribute on the model if not specified; which would be the first non-primary-key column in the db, if not otherwise specified;
        'itemsPerPage' => 10, // falls back to 10 if not specified
    ]
];

// so in short form it could be:
protected $fetch = [
    'category' => [
        'model' => Backpack\NewsCRUD\app\Models\Category::class,
        'searchableAttributes' => ['name'],
    ]
];

// but if identifiableAttribute is introduced most cases won't even need to specify searchableAttributes
// so there's no reason we can't allow this array to contain just classes, instead of definition arrays
// so this could be the shortest form: 
protected $fetch = [
    'category' => Backpack\NewsCRUD\app\Models\Category::class,
];
```

PROs:
- fixes (Problem 1) in Solution 1; you'll be able to have multiple endpoints, that respond to different stuff, search within different attributes etc.;
- fixes (Problem 2) in Solution 1; you'd be able to create selects that use Models that don't have CRUDs; 
- BONUS: you'd only need to change ```MonsterCrudController``` when adding a select field to ```MonsterCrudController```; so all needed changes are in this one file; this would make it much easier to add selects with AJAX;

CONs:
- this would expose 2 new routes for each ```select2_from_ajax``` field you create; if you use this field a lot, your ```php artisan route:list``` could become bloated;
-  it seems like we've polluted the MonsterCrudController (and its routes) with logic about Category; inside the ```project.com/admin/moster``` route domain we'd now have 
     - ```project.com/admin/monster/fetch/category```
     - ```project.com/admin/monster/fetch/category/{id}```
 it makes sense, in a way, since they're only used there; but in another way, it feels like bloat...

I can't say I'm 100% happy with either of the solutions above. Nor am I happy with the naming, but we'll figure that one out. But right now, to me, it sounds like Solution 2 is better. Sure it might lead to bloated routes, but it's much faster to write, and pretty easy to understand.

I'm open to opinions and/or better solutions.

Cheers!

**Comment by tabacitu** [comment]
@pxpm I'm afraid I didn't understand from your reply here https://github.com/Laravel-Backpack/CRUD/pull/2280#issuecomment-562858466 - what's your take on this? Which one do you think would be better?

**Comment by tabacitu** [comment]
depends on #2050

**Comment by tabacitu** [comment]
Let's close the issue - we have a complete & working solution in #2308 , which will get merged into Backpack 4.1 in 2-3 weeks. Plus, it's named ```Fetch``` which... is a lot better, I think 😆
